common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -1267,6 +1268,128 @@ public static UTF8String[] icuSplitSQL(final UTF8String string, final UTF8String
+            CodePointIteratorType.CODE_POINT_ITERATOR_MAKE_VALID);
+
+    // Building the title cased target with 'sb'.
+    StringBuilder sb = new StringBuilder();
+    // 'newWord' is true if the current character is the beginning of a word, false otherwise.
+    boolean newWord = true;
+    // We are maintaining if the current character is preceded by a cased letter.
+    // This is used when lowercasing capital Greek letter sigma ('Σ'), to figure out if it should be
+    // lowercased into σ or ς.
+    boolean precededByCasedLetter = false;
+
+    // 'offset' is a byte offset in target's byte array pointing to the beginning of the character
+    // that we need to process next(this is only actually used in appendLowerCasedGreekCapitalSigma)
+    int offset = 0;
+
+    while(codepointIterator.hasNext()) {
+      int codepoint = codepointIterator.next();
+      // Appending the correctly cased character onto 'sb'.
+      appendTitleCasedCodepoint(sb, codepoint, newWord, precededByCasedLetter, target, offset);
+      // Updating 'newWord', 'precededByCasedLetter' and 'offset' to be ready for the next character
+      // that we will process.
+      newWord = (codepoint == ASCII_SPACE_CODEPOINT);
+      if(!UCharacter.hasBinaryProperty(codepoint, UProperty.CASE_IGNORABLE)) {
+        precededByCasedLetter = UCharacter.hasBinaryProperty(codepoint, UProperty.CASED);
+      }
+      offset += UTF8String.numBytesForFirstByte(target.getByte(offset));
+    }
+    return UTF8String.fromString(sb.toString());
+  }
+
+  private static void appendTitleCasedCodepoint(
+      StringBuilder sb,
+      int codepoint,
+      boolean isAfterAsciiSpace,
+      boolean precededByCasedLetter,
+      UTF8String target,
+      int offset) {
+    if(isAfterAsciiSpace) {
+      // Titlecasing a character if it is in the beginning of a new word.
+      sb.append(codepointToTitleString(codepoint));
+      return;
+    }
+    if(codepoint == CAPITAL_SIGMA) {
+      // Handling capital Greek letter sigma ('Σ').
+      appendLowerCasedGreekCapitalSigma(sb, precededByCasedLetter, target, offset);
+      return;
+    }
+    // If it's not the beginning of a word, or a capital Greek letter sigma ('Σ'), we lowercase the
+    // character.
+    sb.append(
+      toLowerCase(UTF8String.fromString(new String(Character.toChars(codepoint)))).toString());
+  }
+
+  private static void appendLowerCasedGreekCapitalSigma(
+      StringBuilder sb,
+      boolean precededByCasedLetter,
+      UTF8String target,
+      int offset) {
+    int codepoint;
+    if (!followedByCasedLetter(target,offset) && precededByCasedLetter) {
+      codepoint = SMALL_FINAL_SIGMA;
+    }
+    else {
+      codepoint = SMALL_NON_FINAL_SIGMA;
+    }
+    sb.appendCodePoint(codepoint);
+  }
+
+  /**
+   * Checks if the character beginning at 'offset'(in 'targets' byte array) is followed by a cased
+   * letter.
+   *
+   * @param target
+   * @param offset
+   * @return
+   */
```

Comment:
this does not look very helpful - either describe the parameters, or just feel free to remove these lines (imo, the header of the doc comment is helpful enough)